Predicate for distributed query is executed on lite-member -> ClassNotFoundException

Hi,

I'm trying to migrate our application from Hazelcast 2.6.10 to 3.7.1. We are using the reintroduced lite-member feature in some of the cluster-nodes (distributed executor service).

When we execute `IMap#entrySet(Predicate ...)` on a full-member (like with 2.6.10), we see that e.g. a `com.hazelcast.map.impl.query.QueryOperation` is executed on each lite-member with 3.7.1. 

There are two issues with this:

1. Since lite-members do not carry data, this is unnecessary and slows down the distributed query
1. The lite-members do not have the particular `Predicate` implementation in their classpath, therefor there are errors like this:

```
06.09.2016 10:28:53 ERROR [hz._hzInstance_1_dev.generic-operation.thread-2] c.h.s.i.o.impl.OperationExecutorImpl              : [10.0.16.165]:12106 [dev] [3.7.1] Failed to process packet: Packet{flags=1, isResponse=false, isOperation=true, isEvent=false, partitionId=-1, conn=Connection[id=1, /10.0.16.165:52216->/10.0.16.165:12101, endpoint=[10.0.16.165]:12101, alive=true, type=MEMBER]} on hz._hzInstance_1_dev.generic-operation.thread-2
com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.ClassNotFoundException: ch.concordia.automiq2.grid.store.core.data.LongLongFromToPredicate
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.read(JavaDefaultSerializers.java:224)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:46)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.readObject(AbstractSerializationService.java:216)
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:600)
	at com.hazelcast.map.impl.query.QueryOperation.readInternal(QueryOperation.java:77)
	at com.hazelcast.spi.Operation.readData(Operation.java:513)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:127)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:52)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:46)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:172)
	at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:307)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:381)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:117)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:102)
Caused by: java.lang.ClassNotFoundException: foo.bar.MyPredicateImpl42
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at com.hazelcast.nio.ClassLoaderUtil.tryLoadClass(ClassLoaderUtil.java:151)
	at com.hazelcast.nio.ClassLoaderUtil.loadClass(ClassLoaderUtil.java:120)
	at com.hazelcast.nio.IOUtil$ClassLoaderAwareObjectInputStream.resolveClass(IOUtil.java:358)
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1613)
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1518)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1774)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:371)
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.read(JavaDefaultSerializers.java:219)
	... 13 common frames omitted
```

Best regards from Switzerland,
Peti

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java
@@ -61,6 +61,7 @@ import java.util.concurrent.Callable;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Future;
 
+import static com.hazelcast.cluster.memberselector.MemberSelectors.DATA_MEMBER_SELECTOR;
 import static com.hazelcast.query.PagingPredicateAccessor.getNearestAnchorEntry;
 import static com.hazelcast.spi.ExecutionService.QUERY_EXECUTOR;
 import static com.hazelcast.spi.properties.GroupProperty.QUERY_PREDICATE_PARALLEL_EVALUATION;
@@ -533,7 +534,7 @@ public class MapQueryEngineImpl implements MapQueryEngine {
     }
 
     protected List<Future<QueryResult>> queryOnMembers(String mapName, Predicate predicate, IterationType iterationType) {
-        Collection<Member> members = clusterService.getMembers();
+        Collection<Member> members = clusterService.getMembers(DATA_MEMBER_SELECTOR);
         List<Future<QueryResult>> futures = new ArrayList<Future<QueryResult>>(members.size());
         for (Member member : members) {
             Operation operation = new QueryOperation(mapName, predicate, iterationType);

--- a/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/operation/PutAllOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/operation/PutAllOperation.java
@@ -33,6 +33,8 @@ import com.hazelcast.spi.partition.IPartitionService;
 import java.io.IOException;
 import java.util.Collection;
 
+import static com.hazelcast.cluster.memberselector.MemberSelectors.DATA_MEMBER_SELECTOR;
+
 /**
  * Puts a set of records to the replicated map.
  */
@@ -73,7 +75,7 @@ public class PutAllOperation extends Operation implements IdentifiedDataSerializ
 
     private void publishReplicationMessage(Data key, Data value, VersionResponsePair response) {
         OperationService operationService = getNodeEngine().getOperationService();
-        Collection<Member> members = getNodeEngine().getClusterService().getMembers();
+        Collection<Member> members = getNodeEngine().getClusterService().getMembers(DATA_MEMBER_SELECTOR);
         for (Member member : members) {
             Address address = member.getAddress();
             if (address.equals(getNodeEngine().getThisAddress())) {

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryAdvancedTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/query/QueryAdvancedTest.java
@@ -24,8 +24,12 @@ import com.hazelcast.core.EntryEvent;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.IMap;
 import com.hazelcast.core.MapStoreAdapter;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.DataSerializable;
 import com.hazelcast.nio.serialization.Portable;
 import com.hazelcast.nio.serialization.PortableFactory;
+import com.hazelcast.query.Predicate;
 import com.hazelcast.query.SampleObjects.Employee;
 import com.hazelcast.query.SampleObjects.PortableEmployee;
 import com.hazelcast.query.SampleObjects.ValueType;
@@ -40,12 +44,14 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import java.io.IOException;
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
 
 import static com.hazelcast.test.TimeConstants.MINUTE;
 import static org.junit.Assert.assertEquals;
@@ -56,6 +62,43 @@ import static org.junit.Assert.assertTrue;
 public class QueryAdvancedTest extends HazelcastTestSupport {
 
     private static final long TIMEOUT_MINUTES = 2 * MINUTE;
+
+    @Test
+    public void testQueryOperationAreNotSentToLiteMembers() {
+        TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(2);
+        HazelcastInstance fullMember = nodeFactory.newHazelcastInstance();
+        HazelcastInstance liteMember = nodeFactory.newHazelcastInstance(new Config().setLiteMember(true));
+        assertClusterSizeEventually(2, fullMember);
+
+        IMap<Integer, Integer> map = fullMember.getMap(randomMapName());
+        DeserializationCountingPredicate predicate = new DeserializationCountingPredicate();
+
+        //initialize all partitions
+        for (int i = 0; i < 5000; i++) {
+            map.put(i, i);
+        }
+
+        map.values(predicate);
+        assertEquals(0, predicate.serilizationCount());
+    }
+
+    public static class DeserializationCountingPredicate implements Predicate, DataSerializable {
+        private static final AtomicInteger counter = new AtomicInteger();
+
+        public boolean apply(Map.Entry mapEntry) {
+            return false;
+        }
+
+        public int serilizationCount() {
+            return counter.get();
+        }
+
+        public void writeData(ObjectDataOutput out) throws IOException { }
+
+        public void readData(ObjectDataInput in) throws IOException {
+            counter.incrementAndGet();
+        }
+    }
 
     @Test(timeout = TIMEOUT_MINUTES)
     @SuppressWarnings("deprecation")

